### PR TITLE
[BUGFIX] Fix run_diagnostics for contrib expectations

### PR DESCRIPTION
**Pull request by talagluck**
Changes proposed in this pull request:
- This fixes a bug in the `check_json_test_result` function whereby the result of `list.sort()`, which is None, was being assigned to a variable, when it should have been `sorted(list)`.
- Apart from modifying from `.sort()` to `sorted()`, this also adds a key to `sorted()` to make the sorting function more robust when it comes to comparing items of different data types.
- As a result of fixing this function, a couple of failing tests came to light that had previously not been caught, so this addresses those as well. In the case of the test for `expect_column_values_to_be_between`, I removed the `unexpected_list` due to conflicts between timestamp parsing between `pandas` and `sqlalchemy`.

Thanks to @manyshapes for surfacing!

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/self_check/util.py [hunk 1]
```diff
@@ -1904,10 +1904,10 @@ def check_json_test_result(test, result, data_asset=None):
...
                 if (isinstance(value, list)) & (len(value) >= 1):
                     if type(value[0].__lt__(value[0])) != type(NotImplemented):
-                        value = value.sort()
-                        result["result"]["unexpected_list"] = result["result"][
-                            "unexpected_list"
-                        ].sort()
+                        value = sorted(value, key=lambda x: str(x))
+                        result["result"]["unexpected_list"] = sorted(
+                            result["result"]["unexpected_list"], key=lambda x: str(x)
+                        )
```
Comment (cdkini) [inline great_expectations/self_check/util.py:1910]: Ah `sort()` vs `sorted()` is always annoying. Great catch!

What are the types of `value` and `unexpected_list` here? Do we need the custom comparator for sorting? I think its perfectly fine but just asking for my knowledge.

File: tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json [hunk 2]
```diff
@@ -197,7 +197,6 @@
         "title": "basic_negative_test_with_timestamps",
         "exact_match_out": false,
         "out": {
-          "unexpected_list": ["Jan 01 2001 12:00:01"],
           "unexpected_index_list": [0, 9],
```
Comment (cdkini) [inline tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json:200]: Could you clarify what this removal is doing exactly?

**Review by cdkini — APPROVED** [review]
I left a few minor comments but none of them are blocking so LGTM 🚀 

This is more of a general software engineering question but if we find bugs that haven't been caught by our test suite, should we add new tests to cover the missing functionality upon fixing the bug? So in the case of `None` being returned by `sort()`, should we add a test that ensures we properly handles cases that failed before the change to `sorted()`?

File: great_expectations/self_check/util.py (hunk 1 above)
Comment (talagluck) [reply great_expectations/self_check/util.py:1910]: The types for both of these should be `list`. We do a check on L1905 to see if `value` is a list, and we only get here if that resolves to True. We also assume that `unexpected_list` is True, and I think that's just a fundamental part of expectations (I'm not sure whether we test than anywhere). I can add some typing here as well - I'm not going to add it everywhere in this module, so it may look a bit funny, but if adding it here makes it clearer, I can do that.

File: tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json (hunk 2 above)
Comment (talagluck) [reply tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json:200]: Essentially, this json test is running a test for this expectation for both pandas and sqlalchemy. I believe that `parse_strings_as_datetimes` functions differently for each of these. So in the case of pandas, the unexpected list would look like what was there (though it was previously missing the 1970 date from L9). However in the case of sqlalchemy, this date actually gets parsed into "2001-01-01 12:00:01" or thereabouts. (I also might be mixing up pandas and sqlalchemy here). Though the expectation for both backgrounds is correct, the unexpected list for each would be slightly different. I could split out the section completely for the two backends, but that adds a fair amount of complexity, and we have most of the functionality due to unexpected_index_list - this tells us the index of  the unexpected dates. So I decided to go with that approach.

File: great_expectations/self_check/util.py (hunk 1 above)
Comment (cdkini) [reply great_expectations/self_check/util.py:1910]: Nope I think that explanation is more than sufficient for me :) thanks for clarifying.

**Comment by talagluck** [comment]
Thanks for the review, @cdkini! To your point about adding test coverage - the way I'm looking at this, we thought we had test coverage here, and we didn't, due to the sorted/sort issue. That is fixed now, and so now we should catch any errors not previously caught (which is the case in the two other files amended here. Adding additional test coverage feels like we would be testing our testing capability. Am I thinking about this right? Happy to chat.

File: tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json (hunk 2 above)
Comment (cdkini) [reply tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json:200]: That makes a lot of sense as well. Do you think we should add a ticket about bifurcating these backends to our backlog? I think that split might be helpful but I'm not sure if the value it'll provide is worth the effort.

File: great_expectations/self_check/util.py (hunk 1 above)
Comment (talagluck) [reply great_expectations/self_check/util.py:1910]: Ah, and the custom comparator is due to an issue that was arising if the values in the list were of different types. So if one of the values was a list, but the rest were strings, this would raise an error due to trying to sort items of different types.

File: tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json (hunk 2 above)
Comment (talagluck) [reply tests/test_definitions/column_map_expectations/expect_column_values_to_be_between.json:200]: My gut tells me that the bifurcation would not be worthwhile. I think it would really only help in very severe edge cases, like this one, and it would introduce a bunch of complexity and duplicated code.

It might be worth a re-visit of this testing infrastructure entirely to make sure that it's effective for V3. When we initially set it up, it was basically a port of V2 functionality to V3. I'm not sure how much additional thought and work has been done to it since then, but it's possible that it could stand some updating.

**Comment by cdkini** [comment]
Hmm I think that makes sense. My question was more general and less specific to this PR so it may have been better to have discussed this externally. Sorry if it's off-topic.

The change you made is a very internal part of the method and so, I think isolating that specific behavior in a test would be difficult. Plus, more tests are not always the answer. I more so asked the question to get a better sense of when and when not to add tests but I think the points you've made are valid and I'm happy to have this merged as is!
